### PR TITLE
Initial implementation of iOS bindings to Lifecycle

### PR DESCRIPTION
**Pull request by elijah-semyonov**
Initial implementation of iOS bindings to Lifecycle.

ios_lifecycle

### CODE REVIEW

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ApplicationStateListener.uikit.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import platform.UIKit.UIApplicationDidEnterBackgroundNotification
+import platform.UIKit.UIApplicationState
+import platform.UIKit.UIApplicationWillEnterForegroundNotification
+import platform.darwin.NSObject
+
+internal class ApplicationStateListener(
+    /**
+     * Callback which will be called with `true` when the app becomes active, and `false` when the app goes background
+     */
+    private val callback: (Boolean) -> Unit
```
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ApplicationStateListener.uikit.kt:32]: TODO: rename

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ios_lifecycle.png
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ios_lifecycle.png:1]: Let's avoid storing such images in git history

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ios_lifecycle.excalidraw
Comment (MatkovIvan) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ios_lifecycle.excalidraw:1]: Let's move it out of our git history too

**Review by MatkovIvan — APPROVED** [review]
🔥

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ApplicationStateListener.uikit.kt (hunk 1 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ApplicationStateListener.uikit.kt:32]: Rename what exactly?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ios_lifecycle.png
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ios_lifecycle.png:1]: Removed

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ios_lifecycle.excalidraw
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ios_lifecycle.excalidraw:1]: Removed

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ApplicationStateListener.uikit.kt (hunk 1 above)
Comment (MatkovIvan) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ApplicationStateListener.uikit.kt:32]: Well, if I knew a specific good-fit name, I'd suggest it)

Comment (ASalavei) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ApplicationStateListener.uikit.kt:32]: `onApplicationActiveStateChanged` ?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ViewControllerBasedLifecycleOwner.uikit.kt [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    fun handleViewDidDisappear() {
+        reduce(Action.VIEW_DID_DISAPPEAR)
+    }
+
+    private fun reduce(action: Action) {
+        val currentState = state
+        state = state.reduce(action)
+        println("$currentState -> $action -> $state")
```
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ViewControllerBasedLifecycleOwner.uikit.kt]: Debug code? Should it go to production?

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ViewControllerBasedLifecycleOwner.uikit.kt [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+    override val lifecycle = LifecycleRegistry(this)
+
+    private var state: State = State.Created(
+        isApplicationForeground = ApplicationStateListener.isApplicationActive,
+        lifecycle = lifecycle
+    )
+
+    private val applicationStateListener = ApplicationStateListener { isForeground ->
+        state = state.reduce(
```
Comment (ASalavei) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ViewControllerBasedLifecycleOwner.uikit.kt]: It's better to call `this.reduce()` here to have single point of state changes.

Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ViewControllerBasedLifecycleOwner.uikit.kt]: That's indeed an oversight, fixed

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ViewControllerBasedLifecycleOwner.uikit.kt (hunk 2 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ViewControllerBasedLifecycleOwner.uikit.kt]: It wasn't going to jb-main, so we'd remove it later during, but I'll do it now.

**Pull request merged**